### PR TITLE
change: check etcd cluster version when init_etcd

### PR DESCRIPTION
**Pull request by tokers**
Signed-off-by: tokers <zchao1995@gmail.com>

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [ ] Is this PR backward compatible?

### CODE REVIEW

**Comment by moonming** [comment]
the etcd in CI is 3.2, so failed: https://github.com/apache/apisix/pull/2233/checks?check_run_id=1116987808#step:6:288

**Comment by tokers** [comment]
@moonming It seems the version selection for etcd is not specified in `./bin/apisix` or other stuffs. Is it under the Github Action's control?

**Comment by moonming** [comment]
we used to be the default version installed by github actions, now it is the etcd 3.4 version installed manually in https://github.com/apache/apisix/pull/2036

**Comment by tokers** [comment]
@moonming OK, let me rebase my branch. By the way, i think it's better to check out the CHANGELOG of etcd, AKAIK there are some bugs in 3.x series releases, but i'm not sure the exact version range.

**Comment by tokers** [comment]
@moonming I found PR  #2036 also checked the etcd version, but IMHO the way is wrong.

Firstly, when we ask the version message from etcd, it returns a JSON string like this:

```json
{"etcdserver":"3.5.0-pre","etcdcluster":"3.5.0"}
```

we should check the version of `etcdcluster`, not the endpoint's server version (`etcdserver`) that we requested, the cluster version is the minimal server version of all endpoints in the cluster. So that's the one we need to check.

Secondly, the semantic version check method doesn't consider the `pre release` situation, although it's rare in prod, i think's better to consider it.

**Comment by moonming** [comment]
nice! we can fix it in this PR :)

Thanks,
Ming Wen
Twitter: _WenMing


Alex Zhang <notifications@github.com> 于2020年9月16日周三 上午10:36写道：

File: bin/apisix [hunk 1]
```diff
@@ -885,20 +939,44 @@ local function init_etcd(show_output)
...
-        cmd = "curl " .. host .. "/version 2>&1"
-        local res = excute_cmd(cmd)
-        local op_ver = str_split(res, "\"")[4]
-        local need_ver = "3.4.0"
-        if not check_version(op_ver, need_ver) then
-            io.stderr:write("etcd version must >=", need_ver, " current ", op_ver, "\n")
+        uri = host .. "/version"
+        local cmd = string.format("curl -s -m %d %s | sed -e 's/[{}]/''/g'", timeout * 2, uri)
+        local res = execute_cmd(cmd)
+        local m = split(res, [[%s*,%s*]])
```
Comment (membphis) [inline bin/apisix:949]: I think we can use `dkjson` the parse the JSON response body, it was written by pure Lua code.

https://luarocks.org/modules/dhkolf/dkjson

File: bin/apisix [hunk 2]
```diff
@@ -885,20 +939,44 @@ local function init_etcd(show_output)
...
     for index, host in ipairs(yaml_conf.etcd.host) do
-        -- check if etcd version above 3.4
-        cmd = "curl " .. host .. "/version 2>&1"
-        local res = excute_cmd(cmd)
-        local op_ver = str_split(res, "\"")[4]
-        local need_ver = "3.4.0"
-        if not check_version(op_ver, need_ver) then
-            io.stderr:write("etcd version must >=", need_ver, " current ", op_ver, "\n")
+        uri = host .. "/version"
+        local cmd = string.format("curl -s -m %d %s | sed -e 's/[{}]/''/g'", timeout * 2, uri)
```
Comment (membphis) [inline bin/apisix:947]: it depends on the tool `sed`. External tools, the less dependency the better.

File: bin/apisix [hunk 3]
```diff
@@ -885,20 +939,44 @@ local function init_etcd(show_output)
...
+            end
+        end
+
+        if not cluster_version then
+            io.stderr:write(errmsg)
             return
         end
 
+        if compare_semantic_version(cluster_version, min_etcd_version) then
+            io.stderr:write(string.format("etcd cluster version \"%s\" is less than the required version \"%s\", please upgrade your etcd cluster", cluster_version, min_etcd_version))
```
Comment (membphis) [inline bin/apisix:971]: this line is too long. it is not easy for reading

File: bin/apisix (hunk 1 above)
Comment (tokers) [reply bin/apisix:949]: That'd be better.

File: bin/apisix (hunk 2 above)
Comment (tokers) [reply bin/apisix:947]: OK, will use jkjson.

File: bin/apisix (hunk 3 above)
Comment (tokers) [reply bin/apisix:971]: OK, will split this line.

**Comment by tokers** [comment]
@membphis Fixed.

**Review by membphis — APPROVED** [review]

**Review by moonming — APPROVED** [review]

**Comment by Yiyiyimu** [comment]
@tokers Thank you for the fix!

**Pull request merged**